src/main/java/com/gotocompany/depot/maxcompute/client/MaxComputeClient.java

```diff
@@ -0,0 +1,155 @@
+import com.gotocompany.depot.maxcompute.client.insert.InsertManager;
+import com.gotocompany.depot.maxcompute.client.insert.InsertManagerFactory;
+import com.gotocompany.depot.maxcompute.exception.MaxComputeTableOperationException;
+import com.gotocompany.depot.maxcompute.model.RecordWrapper;
+import com.gotocompany.depot.maxcompute.schema.SchemaDifferenceUtils;
+import com.gotocompany.depot.metrics.Instrumentation;
+import com.gotocompany.depot.metrics.MaxComputeMetrics;
+import lombok.AllArgsConstructor;
+import lombok.NoArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+@AllArgsConstructor
+@NoArgsConstructor
+@Slf4j
+public class MaxComputeClient {
+
+    private Odps odps;
+    private MaxComputeSinkConfig maxComputeSinkConfig;
+    private TableTunnel tableTunnel;
+    private InsertManager insertManager;
+    private MaxComputeMetrics maxComputeMetrics;
+    private Instrumentation instrumentation;
+
+    public MaxComputeClient(MaxComputeSinkConfig maxComputeSinkConfig,
+                            Instrumentation instrumentation, MaxComputeMetrics maxComputeMetrics) {
+        this.maxComputeSinkConfig = maxComputeSinkConfig;
+        this.instrumentation = instrumentation;
+        this.odps = initializeOdps();
+        this.tableTunnel = new TableTunnel(odps);
+        this.tableTunnel.setEndpoint(maxComputeSinkConfig.getMaxComputeTunnelUrl());
+        this.maxComputeMetrics = maxComputeMetrics;
+        this.insertManager = initializeInsertManager();
+    }
+
+    public void upsertTable(TableSchema tableSchema) throws OdpsException {
+        String projectName = maxComputeSinkConfig.getMaxComputeProjectId();
+        String datasetName = maxComputeSinkConfig.getMaxComputeSchema();
+        String tableName = maxComputeSinkConfig.getMaxComputeTableName();
+        if (!this.odps.tables().exists(tableName)) {
+            createTable(tableSchema, projectName, datasetName, tableName);
+            return;
+        }
+        Instant start = Instant.now();
+        updateTable(tableSchema, projectName, datasetName, tableName);
+        instrumentation.logInfo("Successfully updated maxCompute table " + tableName);
+        instrument(start, MaxComputeMetrics.MaxComputeAPIType.TABLE_UPDATE);
+    }
+
+    private void createTable(TableSchema tableSchema, String projectName, String datasetName, String tableName) throws OdpsException {
+        Instant start = Instant.now();
+        this.odps.tables().create(projectName, datasetName, tableName, tableSchema, "",
+                false, maxComputeSinkConfig.getMaxComputeTableLifecycleDays(),
+                null, null);
+        instrumentation.logInfo("Successfully created maxCompute table " + tableName);
+        instrument(start, MaxComputeMetrics.MaxComputeAPIType.TABLE_CREATE);
+    }
```

Comment:
The table creation doesn't handle concurrent creation attempts and lacks proper retry mechanism for transient failures. Also missing proper validation of table parameters.

Reply:
retry added, concurrent creation should've been catered with synchronzation on depot listener part